blocks/basic/include/gnuradio-4.0/basic/FunctionGenerator.hpp

```diff
@@ -0,0 +1,297 @@
+
+* ParabolicRamp
+  - Produces a function with three sections: parabolic-linear-parabolic. The 'roundOffTime' parameter determines the length of the parabolic sections.
+  To create a `property_map` use `createParabolicRampPropertyMap(tagIndex, startValue, finalValue, duration, roundOffTime)`.
+
+* CubicSpline
+  - Implements smooth cubic spline interpolation between start and final values over a specified duration.
+  To create a `property_map` use `createCubicSplinePropertyMap(tagIndex, startValue, finalValue, duration)`.
+
+* ImpulseResponse
+  - Creates a function that spikes from 'time0' for a duration of 'time1'.
+  To create a `property_map` use `createImpulseResponsePropertyMap(tagIndex, startValue, finalValue, time0, time1)`.
+
+To create a chain of functions one can use `ClockSource`. There are 2 ways to do it:
+1) Using times + property_map
+Firstly, add time Tag entries, with time and command string:
+@code
+auto &clockSrc = testGraph.emplaceBlock<gr::basic::ClockSource<float>>({ { "sample_rate", 100.f }, { "n_samples_max", 100 }, { "name", "ClockSource" } });
+clockSrc.addTimeTagEntry(1'000, "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=1");
+clockSrc.addTimeTagEntry(10'000, "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=2");
+clockSrc.addTimeTagEntry(30'000, "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=3");
+@endcode
+
+Secondly, create table to match command and functions:
+@code
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=1" } }, createConstPropertyMap(5.f));
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=2" } }, createLinearRampPropertyMap(5.f, 30.f, .2f));
+funcGen.addFunctionTableEntry({ { "context", "CMD_BP_START:FAIR.SELECTOR.C=1:S=1:P=3" } }, createConstPropertyMap(30.f));
+@endcode
+
+Lastly create a matching predicate with the following signature:
+@code
+std::function<std::optional<bool>(const property_map &, const property_map &, std::size_t)>;
+@endcode
+
+
+2) Using user defined ready-to-use Tags.
+@code
+auto &clockSrc = testGraph.emplaceBlock<gr::basic::ClockSource<float>>({ { "sample_rate", 100.f }, { "n_samples_max", 100 }, { "name", "ClockSource" } });
+clockSrc.tags = { Tag(0, createConstPropertyMap(5.f)),
+                  Tag(100, createLinearRampPropertyMap(5.f, 30.f, .2f)),
+                  Tag(300, createConstPropertyMap(30.f)),
+                  Tag(350, createParabolicRampPropertyMap(30.f, 20.f, .1f, 0.02f)),
+                  Tag(550, createConstPropertyMap(20.f)),
+                  Tag(650, createCubicSplinePropertyMap(20.f, 10.f, .1f)),
+                  Tag(800, createConstPropertyMap(10.f)),
+                  Tag(850, createImpulseResponsePropertyMap(10.f, 20.f, .02f, .06f)) };
+@endcode
+
+)"">;
+
+template<typename T>
+    requires(std::floating_point<T>)
+struct FunctionGenerator : public gr::Block<FunctionGenerator<T>, BlockingIO<true>, FunctionGeneratorDoc> {
+    /**
+     * TODO: Taken from CtxSettings
```

Comment:
fyi: need to follow-up this 'TODO' comment. Will do it in my branch since I also needed the tag-matching-predicate more intensely.